cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -392,10 +390,82 @@ def _validate_feature_id(self, feature_id: str, df_name: str):
 
         return
 
+    def _get_matrix_format(self, matrix: Union[np.ndarray, sparse.spmatrix]) -> str:
+        """
+        Given a matrix, returns the format as one of: csc, csr, coo, dense
+        or unknown.
+
+        This mimics the scipy.sparse `format` property, but extends it to
+        support ndarray and other classes AnnData may proxy the matrix with.
+        """
+
+        # Note: the AnnData proxy classes DO support the `format_str` property, but
+        # doing a slice seemed safer, if less performant.  Using `format_str`, which
+        # currently works, uses private API:
+        #
+        # >>> return getattr(matrix, "format_str", "dense)
+        #
+        format = "unknown"
+        if self.adata.n_obs == 0 or self.adata.n_vars == 0:
+            format = "dense"
+        else:
+            matrix_slice = matrix[0:1, 0:1]
```

Comment:
Question: why do we need this slice to infer the type of the matrix?

Reply:
Because the type of X when you open in backed is a private AnnData-internal class which has no public API to introspect it (and the type itself is private). If we hard-wire the private (internal) type system or use other private interfaces, it is subject to breaking.

Using the public __getitem__ is guaranteed to return it in a form that supports public type introspection.